src/volatile_set.c

```diff
@@ -1,79 +1,1162 @@
 #include <string.h>
 #include "volatile_set.h"
+#include "rax.h"
 #include "zmalloc.h"
-#include "config.h"
 #include "endianconv.h"
 #include "serverassert.h"
+#include "hashtable.h"
+#include "server.h"
+#include <stdint.h>
 
-#define EXPIRY_HASH_SIZE 16
-volatile_set *createVolatileSet(volatileEntryType *type) {
-    volatile_set *set = zmalloc(sizeof(volatile_set));
-    set->etypr = type;
-    set->expiry_buckets = raxNew();
-    return set;
+/*************************************************************************************************************
+ *                                pointer_vector Implementation
+ *************************************************************************************************************/
+
+#define PV_CARD_BITS 30
+#define PV_ALLOC_BITS 34
+#define PV_MAX_ELEMENTS ((1ULL << PV_CARD_BITS) - 1)
+#define PV_HEADER_SIZE (sizeof(pointer_vector))
+#define PV_ELEM_SIZE (sizeof(void *))
+#define PV_ALLOC(pv) (pv ? pv->alloc : 0)
+#define PV_LEN(pv) (pv ? pv->len : 0)
+#define PV_USED_SIZE(pv) (pv ? (PV_HEADER_SIZE + (pv_len(pv)) * PV_ELEM_SIZE) : 0)
+
+/* Custom vector structure with embedded allocation and length counters */
+typedef struct {
+    uint64_t len : 30;   /* Number of elements */
+    uint64_t alloc : 34; /* Allocated capacity */
```

Comment:
It's not obvious that the capacity is in bytes. I would have assumed it's the number of elements that can fit in the allocated array.

Are there any benefits to storing it in bytes?

Reply:
we use both. the alloc is used in order to avoid call to realloc if not needed and the len is used for tracking cardinality. this is the same as done for listpack